### PR TITLE
run tests if 'composer test' is defined

### PR DESCRIPTION
**Pull request by dj95**
The buildpack for php should feature tests, like in the nodejs or ruby buildpack. As the nodejs features tests with `yarn test` or `npm test`the php-buildpack should feature tests run
with `composer test`.

This pull request contains a small test script, that executes the test command if it is definded.
If the command is not defined, a notice is printed.

### CODE REVIEW

**Comment by Soolan** [comment]
Hi there,
Any update on this?
Tx

**Comment by Soolan** [comment]
My pleasure @dj95 
It looks good to me, what I don't understand is why a simple PR like this takes forever to be merged.
We are still waiting for this issue to be resolved.

**Comment by dzuelke** [comment]
@Soolan it's not up to you to approve PRs on a repository you do not own. I unfortunately cannot remove your approval as the GitHub UI won't let me.

Furthermore, it's not an issue that's blocking anybody in any way. You can put the `test` command into `app.json`, and everything will work just fine.

File: bin/test [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+main() {
+    # parse the arguments into variables
```
Comment (dzuelke) [inline bin/test]: This comment is incorrect; it doesn't parse multiple arguments into variables.

File: bin/test [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    declare build_dir="${1:-}"
```
Comment (dzuelke) [inline bin/test]: Why `declare`? You meant `local`?

Comment (dzuelke) [reply bin/test]: `bin/test` doesn't receive arguments, so this whole bit is redundant anyway.

File: bin/test [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+    # check if composer tests are available
+    if composer run-script -l 2>/dev/null | grep -q '^\s*test\b'; then
+        # cd into the build directory and run the tests via composer
+        cd "$build_dir" && composer test
```
Comment (dzuelke) [inline bin/test]: You could `cd ${1:-}` here directly, but see above

Comment (dzuelke) [reply bin/test]: You probably want `composer run-script --timeout=0 test`

File: bin/test [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    else # if no tests are defined...
+        # ...display a friendly reminder to define some
+        printf "\033[0;31mNo test command found! Please define a test command in the composer.json in order to run test.\033[0m\n"
```
Comment (dzuelke) [inline bin/test]: It needs to be discussed whether this should result in an error or not, given how users have to enable CI explicitly.

File: bin/test [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    fi
+}
+
+# fail hard
+set -eo pipefail
+# fail harder
+set -eu
+
+# execute the main function with command line arguments
+main "$@"
```
Comment (dzuelke) [inline bin/test]: Please move the entire logic out of that function.

File: bin/test (hunk 2 above)
Comment (josegonzalez) [reply bin/test]: Looks like it does, based on the [nodejs buildpack](https://github.com/heroku/heroku-buildpack-nodejs/blob/master/bin/test).

File: bin/test (hunk 3 above)
Comment (josegonzalez) [reply bin/test]: I think the `cd` is correct, as noted above (and [here](https://github.com/heroku/heroku-buildpack-nodejs/blob/master/bin/test#L5)).

File: bin/test (hunk 4 above)
Comment (josegonzalez) [reply bin/test]: The `nodejs` buildpack doesn't check if there is a test command, so maybe this buildpack shouldn't either.

Comment (dzuelke) [reply bin/test]: But that results in `npm` printing an error, yet still exiting with code 0. Other languages behave differently. And the underlying Heroku CI system will error the run (not fail it) if no tests are defined (and the buildpack doesn't have default test run behavior).

**Comment by Soolan** [comment]
@dzuelke 
hmmmm let me see... it is not up to me yet I did it ;)
Sweetheart, go to settings and restrict outsider from approving. Believe me, it will look more professional than attacking ppl for approval =))

File: bin/test (hunk 2 above)
Comment (josegonzalez) [reply bin/test]: I was responding to this bin receiving arguments, not the declare vs local vs not assigning.

File: bin/test (hunk 4 above)
Comment (josegonzalez) [reply bin/test]: I think its worth just relying on the underlying toolchain. Its fairly easy for a user to set their test command to `true` if they decide that they do not want tests from their php buildpack to be invoked.

**Comment by dzuelke** [comment]
The problem is not your approval, @Soolan (our internal compliance status check tool filters that), but the 

> It looks good to me, what I don't understand is why a simple PR like this takes forever to be merged.

comment.

And I'm not your sweetheart.

File: bin/test (hunk 2 above)
Comment (dj95) [reply bin/test]: I referred to the nodejs buildpack, like @josegonzalez wrote. The idea behind this declaration is that everyone who looks at the code exactly knows what the first argument of this function does. It's just for better readability.

`declare`is used in here, as `local`shifts the arguments. E.g.

```
main() {
    local arg1=$1 # would be $1 with declare
    local arg2=$1 # would be $2 with declare
}
```

File: bin/test (hunk 4 above)
Comment (dj95) [reply bin/test]: The check for the `composer test` command should be in done in order to not break existing CI pipelines. This warning may be irrelevant, but notifies the user to be able to test with `composer test`. But I can remove it, if you don't like the notification.

File: bin/test (hunk 5 above)
Comment (dj95) [reply bin/test]: I've done it in c7c79064f5da75e4a4c220520f9583132f4c0647

File: bin/test (hunk 2 above)
Comment (josegonzalez) [reply bin/test]: @dj95 for consistency's sake, lets just move to not having a function? I don't think thats the way any of the buildpacks are structured for bins (unfortunately) so introducing a new pattern isn't great :(

We can just add a comment at the top with a list of args etc. this thing is expecting.

Comment (dj95) [reply bin/test]: I removed the function, as you can see in the new commits, and moved the declare statement to the top in order to keep the readability :)

**Comment by dj95** [comment]
Do you have any updates on the recent commits? :)

File: bin/test [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+set -eo pipefail
+# fail harder
+set -eu
+
+
+# name the command line argument for this script
+declare build_dir="${1:-}"
+
+# check if composer tests are available
+if cd "$build_dir" && composer run-script -l 2>/dev/null | grep -q '^\s*test\b'; then
```
Comment (josegonzalez) [inline bin/test]: This could just be a separate:

```
pushd "$build_dir" > /dev/null
```

So you don't have to cd in the if statement.

Comment (dj95) [reply bin/test]: Thank you for this hint! I think it's better to leave the pushd in the if-statement in order to check whether the script runs the test-command in an existing directory or not.

**Comment by coudenysj** [comment]
Are there any plans on adding this (or an alternative) to the buildpack?

**Comment by dj95** [comment]
I'm still waiting for more feedback in order to bring this pull request into a state that is acceptable for merging. If I get feedback, I'll implement it as soon as possible.

**Comment by tortuetorche** [comment]
Hi @dj95,

Any news on this pull request?
This will be really useful with GitLab Auto DevOps feature, see https://gitlab.com/gitlab-org/gitlab-ce/issues/47651

Have a good day,
Tortue Torche

File: bin/test (hunk 4 above)
Comment (scasei) [reply bin/test]: if that error messages prevents this pr from beeing merged, than just remove it, please @dj95

**Comment by scasei** [comment]
@dzuelke Last commit has closed the open discussion, what is left to get this PR merged?

**Comment by dj95** [comment]
It's quite frustrating to see, how the maintainers care about feature requests and pull requests for this buildpack from the community. I mean they even didn't manage to get an updated logo merged for almost 2 years (https://github.com/heroku/heroku-buildpack-php/pull/154 - and this pull requests is from a heroku itself). Since there's no feedback or a sign that they will merge it, even if the code is well documented and only 20 lines long, I might close this issue in a few weeks as I don't see the point to offer this contribution when the maintainers do not want pull requests from the community.

**Comment by Soolan** [comment]
@dj95 I can not agree more with you on this issue. This is just an example of irresponsible behavior from the repo maintainers.

**Comment by dzuelke** [comment]
Please see #350 for a merged implementation.

**Pull request closed without merging**